shell/src/context_listener.rs

```diff
@@ -317,3 +250,83 @@ fn listen_protocol_events(
 
     Ok(())
 }
+
+fn perform_context_action(
+    action: &ContextAction,
+    context: &mut Box<dyn ContextApi>,
+) -> Result<(), Error> {
+    match action {
+        ContextAction::Get { key, .. } => {
+            context.get_key(key)?;
+        }
+        ContextAction::Mem { key, .. } => {
+            context.mem(key)?;
+        }
+        ContextAction::DirMem { key, .. } => {
+            context.dirmem(key)?;
+        }
+        ContextAction::Set {
+            key,
+            value,
+            context_hash,
+            ..
+        } => {
+            context.set(context_hash, key, value)?;
+        }
+        ContextAction::Copy {
+            to_key: key,
+            from_key,
+            context_hash,
+            ..
+        } => {
+            context.copy_to_diff(context_hash, from_key, key)?;
+        }
+        ContextAction::Delete {
+            key, context_hash, ..
+        } => {
+            context.delete_to_diff(context_hash, key)?;
+        }
+        ContextAction::RemoveRecursively {
+            key, context_hash, ..
+        } => {
+            context.remove_recursively_to_diff(context_hash, key)?;
+        }
+        ContextAction::Commit {
+            parent_context_hash,
+            new_context_hash,
+            block_hash: Some(block_hash),
+            author,
+            message,
+            date,
+            ..
+        } => {
+            let hash = context.commit(
+                block_hash,
+                parent_context_hash,
+                author.to_string(),
+                message.to_string(),
+                *date,
+            )?;
+            assert_eq!(
+                &hash,
+                new_context_hash,
+                "Invalid context_hash for block: {}, expected: {}, but was: {}",
+                HashType::BlockHash.hash_to_b58check(block_hash),
+                HashType::ContextHash.hash_to_b58check(new_context_hash),
+                HashType::ContextHash.hash_to_b58check(&hash),
+            );
+        }
+
+        ContextAction::Checkout { context_hash, .. } => {
+            context.checkout(context_hash)?;
+        }
+
+        ContextAction::Commit { .. } => (), // Ignored (no block_hash)
```

Comment:
@bkontur on which situations can this happen? (this was implicit in the old version, but I removed the wildcard match, which I think we should avoid as much as possible).

Reply:
hmm, I dont know, it should not happen, even commit_genesis should be with block_hash:
```
Option.iter
    (fun inspector ->
      inspector#mark_header_and_protocol chain_genesis.block chain_genesis.protocol)
    context_raw_inspector ;
```